aten/src/ATen/TensorIterator.cpp

```diff
@@ -414,9 +464,9 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
       // NB: we skip this on is_meta_, because the temporary allocation here is
       // unnecessary if we aren't going to actually do the compute
       if (config.cast_common_dtype_to_outputs_ && op.is_output && op.current_dtype != common_dtype_ && !is_meta_) {
-        TORCH_INTERNAL_ASSERT(op.tensor->defined());
+        TORCH_INTERNAL_ASSERT(op.tensor_base().defined());
         // Marker [Output original_tensor is set]
-        op.original_tensor = op.tensor;
+        op.original_tensor(op.unsafeReleaseTensor());
```

Comment:
Use of unsafe is suspicious. Is there not a safe way to do this?

Reply:
There is a safe way, but it involves refcount bumps:
```
op.original_tensor(c10::MaybeOwned<TensorBase>::owned(op.tensor_base()))
```

Reply:
a small argument why it's correct would be a big help. I don't know what `original_tensor` takes as argument locally here

Reply:
Actually, why don't you make `std::move(op).tensor_base()` work, which moves the tensor base out of it?

Reply:
>Actually, why don't you make `std::move(op).tensor_base()` work, which moves the tensor base out of it?

This doesn't work because `tensor_base()` returns a reference to `TensorBase`, while `tensor_base_` is a `MaybeOwned<TensorBase>` so we would lose the borrowed status.

Looking at the again, there are only ever two things done with the unsafe ops: exchanging tensor into original_tensor, or moving original_tensor back to tensor. So instead of having the unsafe methods, I've just made these dedicated functions of `OperandInfo`.